### PR TITLE
[READY]Basically, Instant Explosions

### PR DESCRIPTION
**Pull request by Iamgoofball**
## About The Pull Request
Extools maptick stuff is in the game. Stolen from https://github.com/BeeStation/BeeStation-Hornet/pull/1119, improves performance. Requires ex-tools on the server, though.

Explosions have been refactored to do the actual exploding in a subsystem.

Credit to goon.

Here's some videos!

https://file.house/UUmD.mp4

## Why It's Good For The Game
Basically instant max-caps now.

## Changelog
:cl: Goonstation Coders, Beestation, Extools devs
refactor: Explosions have been heavily optimized.
/:cl:

### CODE REVIEW

File: code/controllers/subsystem.dm [hunk 1]
```diff
@@ -1,88 +1,91 @@
...
-	
+	var/static/list/failure_strikes
+
 	/// Next subsystem in the queue of subsystems to run this tick
 	var/datum/controller/subsystem/queue_next
 	/// Previous subsystem in the queue of subsystems to run this tick
 	var/datum/controller/subsystem/queue_prev
 
+	// Does this subsystem process while explosions are going?
+	var/wait_for_explosions = FALSE
```
Comment (optimumtact) [inline code/controllers/subsystem.dm]: you should just use flags_1 (I think this is at the datum level)

and SS_PAUSEABLE define flag (or maybe SS_INTERRUPTABLE)

File: code/controllers/subsystem.dm [hunk 2]
```diff
@@ -98,7 +101,11 @@
 	SHOULD_NOT_OVERRIDE(TRUE)
 	set waitfor = 0
 	. = SS_SLEEPING
-	fire(resumed)
+	if(!wait_for_explosions)
+		fire(resumed)
+	else
+		if(!SSexplosions.is_exploding())
```
Comment (optimumtact) [inline code/controllers/subsystem.dm]: should be a flag too imo something like MC_CONTROL_INTERRUPTED

File: code/controllers/subsystem/explosions.dm [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+					if(SSEX_MED)
+						src.currentrun = medobj.Copy()
+					if(SSEX_HIGH)
+						src.currentrun = highobj.Copy()
+	//cache for sanic speed (lists are references anyways)
+	var/list/currentrun = src.currentrun
+	while(currentrun.len)
+		var/turf/thing = currentrun[currentrun.len]
+		currentrun.len--
+		if(thing)
```
Comment (optimumtact) [inline code/controllers/subsystem/explosions.dm]: should probably be if(!QDELETED(thing))

**Comment by redta1l** [comment]
i want to see the effects of a station nuke 👀

**Comment by XDTM** [comment]
i don't really know what ex-tools is, will it require any particular installations or adjustments when testing locally?

**Comment by Iamgoofball** [comment]
https://github.com/MCHSL/extools

**Comment by MCHSL** [comment]
When testing locally without extools installed, maptick defaults to 20% internal tick usage which is what we currently use

**Comment by stylemistake** [comment]
I wonder what else this maptick extension can be used for. Good work.

**Comment by Iamgoofball** [comment]
oh that shit aint my work, its extools and beestation like the OP says

File: code/__DEFINES/subsystems.dm [hunk 4]
```diff
@@ -165,6 +166,7 @@
 #define FIRE_PRIORITY_ATMOS_ADJACENCY	300
 #define FIRE_PRIORITY_CHAT			400
 #define FIRE_PRIORITY_OVERLAYS		500
+#define FIRE_PRIORITY_EXPLOSIONS	666
```
Comment (Fikou) [inline code/__DEFINES/subsystems.dm]: these arent tabbed like the other ones and it makes me angry

Comment (stylemistake) [reply code/__DEFINES/subsystems.dm]: Tabs for indentation, spaces for alignment.

**Comment by davethwave** [comment]
The day of outrunning explosions are over.

**Comment by GravitronBlue** [comment]
It's going to be the day of Cuban Pete and lots of "I DED" the moment explosions are unavoidable.

This touches everything from TTVs, syndibombs, explosive lances, missiles, and grenades.
There's effectively no way to really guard against explosions either due to the ineffectiveness of bomb armor aside from mechs, so I really hope there's some future refactors planned if this route is being taken.

Assuming so I'll likely start calling for the removal of the toxins mixing lab and TTVs in favor of them being replaced by a bomb payload lab. (Though it's not like scientists wouldn't just switch to making bluespace beaker grenades to avert this, but at least 5-10 minute maxcaps would see a nerf)

> (Lost link to the SS13 arcade remade using HTML/CSS site goes here)

**Comment by optimumtact** [comment]
can you save the soapboxing for the NTR hut

File: code/controllers/subsystem.dm (hunk 2 above)
Comment (actioninja) [reply code/controllers/subsystem.dm]: maybe even roll it into the fire priority system, allow pausing priority downwards or something. Or an independent system for multiple pause priorities.

File: code/controllers/subsystem/air.dm [hunk 5]
```diff
@@ -68,6 +72,15 @@ SUBSYSTEM_DEF(air)
 /datum/controller/subsystem/air/fire(resumed = 0)
 	var/timer = TICK_USAGE_REAL
 
+	if(pipenets_needing_rebuilt.len) // always do this nonsense first
+		var/oldpart = currentpart
+		rebuild_pipenets(resumed)
```
Comment (actioninja) [inline code/controllers/subsystem/air.dm]: I would unwrap this proc, it's called a ton

File: code/controllers/subsystem/explosions.dm [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+			switch(level)
+				if(SSEX_LOW)
+					lowobj.Remove(atom_to_remove)
+				if(SSEX_MED)
+					medobj.Remove(atom_to_remove)
+				if(SSEX_HIGH)
+					highobj.Remove(atom_to_remove)
+
+
+/datum/controller/subsystem/explosions/proc/process_explosions(ex_type, level, resumed = 0)
```
Comment (actioninja) [inline code/controllers/subsystem/explosions.dm]: same for this

File: code/controllers/subsystem/explosions.dm [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+			explode_thing(thing, ex_type, level)
+			remove_from_lists(thing, ex_type, level)
+		else
+			remove_from_lists(thing, ex_type, level)
+		if(ex_type == SSEX_OBJ)
+			if(MC_TICK_CHECK)
+				return
+
+
+/datum/controller/subsystem/explosions/proc/explode_thing(atom_to_explode, ex_type, level)
```
Comment (actioninja) [inline code/controllers/subsystem/explosions.dm]: and this

File: code/controllers/subsystem/explosions.dm [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+		cost_lowobj = MC_AVERAGE(cost_lowobj, TICK_DELTA_TO_MS(TICK_USAGE_REAL - timer))
+
+
+		if(state != SS_RUNNING)
+			return
+		resumed = 0
+		currentpart = SSEXPLOSIONS_TURFS
+	currentpart = SSEXPLOSIONS_TURFS
+
+/datum/controller/subsystem/explosions/proc/remove_from_lists(atom_to_remove, type_of_atom, level)
```
Comment (actioninja) [inline code/controllers/subsystem/explosions.dm]: unwrap here as well

File: code/controllers/subsystem/explosions.dm [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+				if(SSEX_HIGH)
+					highturf.Remove(atom_to_remove)
+		if(SSEX_OBJ)
+			switch(level)
+				if(SSEX_LOW)
+					lowobj.Remove(atom_to_remove)
+				if(SSEX_MED)
+					medobj.Remove(atom_to_remove)
+				if(SSEX_HIGH)
+					highobj.Remove(atom_to_remove)
```
Comment (actioninja) [inline code/controllers/subsystem/explosions.dm]: this feels dumb but I'm not sure if you could do it better. Pass the list directly maybe?

**Review by actioninja — CHANGES_REQUESTED** [review]

**Review by actioninja — CHANGES_REQUESTED** [review]
You should split the extools port into a different PR, it doesn't really seem related to the changes here. Unless there's something that I'm missing.

**Comment by optimumtact** [comment]
it's mostly important to give byond a proper amount of time to do map updates

**Comment by Spepbuthewranglesgithub** [comment]
bombs, and not death waves, i like.

File: code/controllers/subsystem/air.dm (hunk 5 above)
Comment (Iamgoofball) [reply code/controllers/subsystem/air.dm]: what do you mean

Comment (actioninja) [reply code/controllers/subsystem/air.dm]: Procs in byond have stupid amounts of overhead. If you're calling a proc really frequently or in a really performance critical situation where doing things *as fast as possible* such as here is a priority taking the contents of a proc and inlining them straight into the code actually has pretty significant performance improvements.
You could turn them into defines as well if you still want to preserve the code organization, but considering they're only used in one place it wouldn't be that meaningful of a difference imo.

**Comment by actioninja** [comment]
Fair enough, ignore the extools split then.

**Comment by TheVekter** [comment]
This just feels... wrong. Like, it's good and we need it, but something about explosions happening instantly feels weird in SS13. I remember when a max cap would bring the server to a screeching halt for a few seconds.

File: code/controllers/subsystem/explosions.dm (hunk 8 above)
Comment (Iamgoofball) [reply code/controllers/subsystem/explosions.dm]: that would result in copypaste

File: code/controllers/subsystem/explosions.dm (hunk 7 above)
Comment (Iamgoofball) [reply code/controllers/subsystem/explosions.dm]: copypaste

File: code/controllers/subsystem/air.dm (hunk 5 above)
Comment (Iamgoofball) [reply code/controllers/subsystem/air.dm]: aight

File: code/__DEFINES/subsystems.dm (hunk 4 above)
Comment (spookydonut) [reply code/__DEFINES/subsystems.dm]: github doesnt always display tabwidths correctly

File: code/controllers/subsystem.dm [hunk 10]
```diff
@@ -1,88 +1,91 @@
 /**
 	* # Subsystem base class
-	* 
+	*
```
Comment (spookydonut) [inline code/controllers/subsystem.dm]: please turn off strip trailing whitespace setting in your IDE

File: code/controllers/subsystem/air.dm [hunk 11]
```diff
@@ -68,6 +72,15 @@ SUBSYSTEM_DEF(air)
 /datum/controller/subsystem/air/fire(resumed = 0)
 	var/timer = TICK_USAGE_REAL
 
+	if(pipenets_needing_rebuilt.len) // always do this nonsense first
```
Comment (spookydonut) [inline code/controllers/subsystem/air.dm]: length() for speeeeeeeeeeeeeed

File: code/controllers/subsystem/explosions.dm (hunk 8 above)
Comment (spookydonut) [reply code/controllers/subsystem/explosions.dm]: copy paste okay when its a hooooooooot proc

File: code/controllers/subsystem/explosions.dm (hunk 7 above)
Comment (spookydonut) [reply code/controllers/subsystem/explosions.dm]: speeeeeeeeeed

Comment (Iamgoofball) [reply code/controllers/subsystem/explosions.dm]: @optimumtact should we be prioritizing speed or cleanliness here

File: code/controllers/subsystem.dm (hunk 10 above)
Comment (actioninja) [reply code/controllers/subsystem.dm]: iirc this is actually an automatic one now in our editorconfig and this is just an older file.

**Comment by spookydonut** [comment]
is it possible to have the extools use put behind a compiler define such that we can turn it off if extools breaks in the future

**Comment by Ty-the-Smonk** [comment]
> This just feels... wrong. Like, it's good and we need it, but something about explosions happening instantly feels weird in SS13. I remember when a max cap would bring the server to a screeching halt for a few seconds.

Welcome to the future, next up more multi z, and cursed pixel movement

**Comment by Shadowflame909** [comment]
@TheVekter 

This is the SS13 equivalent of inventing the nuke lol.

I hope goof looks upon his creation with more of a "What have I done" feeling then an "I've done it." Because this....This is gonna make things scary, but incredibly realistic. So once more props to goof.

**Comment by optimumtact** [comment]
@spookydonut the only place extools touches is the TICK define, you could have a compiler time define controlling it in compile options I suppose, but if extools goes out of support, the answer is just to remove it from that TICK.

**Comment by ATH1909** [comment]
Aw, but I liked being able to outrun explosions like a total badass...

It made them less bullshit to die to, too, since you could actually potentially avoid taking damage if you were able to react quickly enough. It felt like you could have done _something_ to avoid that death, which made it feel less "cheap".

I will admit that being able to drag a bomb around behind you and not take any damage from it when it explodes was kind of dumb, but it was a fun kind of dumb...

**Comment by optimumtact** [comment]
Please stop complaining about not being able to outrun explosions any more, it's not useful feedback

**Comment by kriskog** [comment]
Testmerge on event hall with loads of meteor spam. Seemed to run very well on around 60 pop, large explosions (I tried a 40/80/160 on CC) will outright freeze the server for a few seconds though which is probably unavoidable but a whole lot better than before.

**Comment by Iamgoofball** [comment]
ran into weird issue with phantom explosions killing people who walk over where they were at before, working on a fix rn

File: code/controllers/subsystem/air.dm (hunk 11 above)
Comment (Iamgoofball) [reply code/controllers/subsystem/air.dm]: are you sure, isnt checking iwth a proc slower than checking the var itself

**Comment by optimumtact** [comment]
can we please lay off the microopts @tgstation/commit-access ? Unless you either bring a profile showing it's faster I don't think you should be making developers jump through hoops on assumptions.

**Comment by optimumtact** [comment]
and I mean profiling the code you're reviewing not some synthetic benchmark from 2 years ago.

**Comment by Cyberboss** [comment]
This contains a critical false assumption that would have been found if it was tested but it obviously hasn't been.

This is also too touchy with the subsystems, try again later if and when extools is initially merged.

**Comment by Iamgoofball** [comment]
>This contains a critical false assumption that would have been found if it was tested but it obviously hasn't been.

What the hell are you talking about?

**Comment by Iamgoofball** [comment]
We literally tested it live. The performance boost was insane. There are videos of me testing it in the OP.

**Comment by optimumtact** [comment]
This is much cleaner than your explosions system cyberboss, you'll have to explain the "assumption"

File: code/controllers/subsystem/air.dm (hunk 11 above)
Comment (nicbn) [reply code/controllers/subsystem/air.dm]: Last time I benchmarked the difference wasn't significant neither consistent enough to favor one over another

**Comment by optimumtact** [comment]
@Cyberboss still waiting

File: code/controllers/subsystem/explosions.dm [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+				O.ex_act(EXPLODE_LIGHT)
+				lowobj.Remove(thing)
+			else
+				lowobj.Remove(thing)
+		cost_lowobj = MC_AVERAGE(cost_lowobj, TICK_DELTA_TO_MS(TICK_USAGE_REAL - timer))
+		if(state != SS_RUNNING)
+			return
+		resumed = 0
+		currentpart = SSEXPLOSIONS_TURFS
+	currentpart = SSEXPLOSIONS_TURFS
```
Comment (Ghilker) [inline code/controllers/subsystem/explosions.dm]: need newline here for The Travis CI build

Comment (Iamgoofball) [reply code/controllers/subsystem/explosions.dm]: right, thanks

File: code/__DEFINES/_tick.dm [hunk 13]
```diff
@@ -1,5 +1,9 @@
+/// Percentage of tick to leave for master controller to run
+#define MAPTICK_MC_MIN_RESERVE 40
+/// internal_tick_usage is updated every tick by extools
+#define MAPTICK_LAST_INTERNAL_TICK_USAGE ((GLOB.internal_tick_usage / world.tick_lag) * 100)
 /// Tick limit while running normally
-#define TICK_LIMIT_RUNNING 80
+#define TICK_LIMIT_RUNNING (max(90 - MAPTICK_LAST_INTERNAL_TICK_USAGE, MAPTICK_MC_MIN_RESERVE))
```
Comment (MrStonedOne) [inline code/__DEFINES/_tick.dm]:
```suggestion
#define TICK_BYOND_RESERVE 10
#define TICK_LIMIT_RUNNING (max(100 - TICK_BYOND_RESERVE - MAPTICK_LAST_INTERNAL_TICK_USAGE, MAPTICK_MC_MIN_RESERVE))
```
or for FUN remove it entirely

File: code/controllers/subsystem.dm (hunk 1 above)
Comment (MrStonedOne) [reply code/controllers/subsystem.dm]: subsystem's already existing flags would be the more appropriate place to put this.

also user definable subsystem vars should go above these in the definition.

File: code/__DEFINES/_tick.dm (hunk 13 above)
Comment (Iamgoofball) [reply code/__DEFINES/_tick.dm]: an interesting idea

**Comment by Iamgoofball** [comment]
We are now good to go.

**Comment by Iamgoofball** [comment]
oh god damn it, didnt mean to axe the TMC tag

**Pull request closed without merging**